autogen/agentchat/contrib/retrieve_user_proxy_agent.py

```diff
@@ -416,21 +534,40 @@ def _generate_retrieve_user_reply(
 
     def retrieve_docs(self, problem: str, n_results: int = 20, search_string: str = ""):
         """Retrieve docs based on the given problem and assign the results to the class property `_results`.
-        In case you want to customize the retrieval process, such as using a different vector db whose APIs are not
-        compatible with chromadb or filter results with metadata, you can override this function. Just keep the current
-        parameters and add your own parameters with default values, and keep the results in below type.
-
-        Type of the results: Dict[str, List[List[Any]]], should have keys "ids" and "documents", "ids" for the ids of
-        the retrieved docs and "documents" for the contents of the retrieved docs. Any other keys are optional. Refer
-        to `chromadb.api.types.QueryResult` as an example.
-            ids: List[string]
-            documents: List[List[string]]
+        The retrieved docs should be type of `QueryResults` which is a list of tuples containing the document and
```

Comment:
at line 511, when the assistant replied back "update context", it is likely to retrieve the doc again with the same query, right now it is calling retrieve_docs() from scratch with a larger return size. For performance consideration, it is desired to have a handler returned from the previous retrieval and iteratively retrieve another set of docs when needed. https://milvus.io/docs/v2.3.x/with_iterators.md 
Not sure we want to take such access pattern into consideration when designing the API for vectorDB.

Reply:
Thank you very much for bringing this up! 
- In the `update context` step, it doesn't always call `retrieve_docs` with the same query, and it could be another issue if we always retrieve a lot of docs which we hope would be enough. Thus, I think the current retrieve n, 3*n, 5*n, ... in the `update context` step should be good. 
- In the db, it would be great to have some kind of cache or iterator, if it can have better performance on this kind of n, 3*n, 5*n, ... iterating retrieval with same query. Applying our case to the milvus iterators, it seems that we need to have a large enough `limit`. Would different `limit` lead to different first 10 results? I'm asking because with ANN search, the results of limit 20 doesn't necessarily contains all the results in limit 10.

Reply:
This is a good question. As you have catched, milvus's iterator requires `limit`, and it is true that different `limit` can sometimes lead to different first 10 results. I don't think there is a good support for iterator/cache in vectorDBs, but if there is need from the upper-layer applications, we would definitely love to work on such features. 
Let me better understand the logic here: you first retrieve n documents, feed [1, x] docs as context to the assistant; when the assistant returns "UPDATE CONTEXT", you use [x+1, 2x] docs from n docs to the assistant? when there is insufficient remaining docs from n docs, you will retrieve 3n from vectorDB? and now which x docs would you feed to the assistant? the top [1, x] from 3n or you will first filter out all n documents from the first retrieval?

Reply:
> and now which x docs would you feed to the assistant? the top [1, x] from 3n or you will first filter out all n documents from the first retrieval?

Check the top [1, 3n] one by one, put into context only if current x not used before, stop while context is long enough or reach n.

Reply:
Hi @siluhuang , would you like to test this PR with your new vector db before approving it?